* **Professor: Hyunsoon Youn**
* **Lecture: Theory and Applications of Deep Learning**
* **Student name: Sreymouy Sao**
* **Student ID: 2021321846**
* **Major: Industrial Engineering, Yonsei University**



**Question 1**




**(a) How many parameters are estimated in this network (including constants)?**

From the table of instance in question (b), we can know that there are 10 nodes in the input layer. From the table in question (c), we can know that there are 12 nodes (classes) in the output layer. So the total number of parameters in the network including the constants (biases) equal (10x50 +50)+(50x40+40)+(40x30+30)+(30x12+12) = 4192

In [ ]:
answer = (10*50 +50)+(50*40+40)+(40*30+30)+(30*12+12) 
print(answer)

4192


(b) **Calculate the value at this node after the sigmoid function is applied.**


In [1]:
import numpy as np 
import math

# weigts
w1 = 0.040065
w2 = 0.246356
w3 = 0.234578
w4 = 0.172821
w5 = 0.449315
w6 = 0.875052
w7 = 0.321868
w8 = 0.141179
w9 = 0.128957
w10 = 0.264436
w = np.array([w1,w2,w3,w4,w5,w6,w7,w8,w9,w10])

# bias
b = 0.031945

# input
x1 = 0.266074
x2 = -0.16562
x3 = 0.32098
x4 = 0.483299
x5 = 0.17234
x6 = 0.273364
x7 = 0.371178
x8 = 0.929823
x9 = 0.251173
x10 = 0.159345
x = np.array([x1,x2,x3,x4,x5,x6,x7,x8,x9,x10])

# sigmoid function 
def sigmoid(x):
  return 1 / (1 + math.exp(-x))


a = sigmoid(np.dot(w,x)+b)
print("The answer is : ",a)

The answer is :  0.6905163466587109


The answer is : 0.6905163466587109

(c) **Calculate the output after a softmax function is applied and estimate the class probabilities for this instance.** 



In [2]:
output_before_softmax = np.array([1.197033, 1.79776, 1.952738, 1.23094, 1.065328, 
                                  0.41491, 1.212204, 1.86248, 1.74077, 1.484271, 
                                  1.950335, 1.657855])

# softmax function
def softmax(x):
  return np.exp(x)/np.sum(np.exp(x))

output_after_softmax = softmax(output_before_softmax)
class_list = np.array(['A', 'B','C','D','E','F','G','H','I','W','X','Y'])
print("Output after softmax function : ", output_after_softmax) 

for i in range(output_after_softmax.size):
  print("The probability of class ", class_list[i], " is ", output_after_softmax[i])


Output after softmax function :  [0.05866384 0.10697023 0.12490189 0.06068706 0.0514247  0.02683483
 0.05956061 0.11412228 0.10104445 0.07818373 0.12460211 0.09300428]
The probability of class  A  is  0.05866384015018051
The probability of class  B  is  0.10697022513283515
The probability of class  C  is  0.12490188710581969
The probability of class  D  is  0.06068706183173598
The probability of class  E  is  0.05142469610033901
The probability of class  F  is  0.026834826118412414
The probability of class  G  is  0.059560614551308344
The probability of class  H  is  0.11412228237783695
The probability of class  I  is  0.10104445069873268
The probability of class  W  is  0.07818372658435027
The probability of class  X  is  0.12460210819971389
The probability of class  Y  is  0.09300428114873514


(d) **For the instance in part (c), what class is assigned?**

In [ ]:
asigned_class = class_list[np.argmax(output_after_softmax)]
print("The assigned class is : ", asigned_class)

The assigned class is :  C


The assigned class is the class that has the maximum probability. It is the class "C".

**Question 3 : For each network below, calculate how many parameters are estimated
(including constants).**

(a)

In [3]:
input_nodes = 128*128
l1_nodes = 50
l2_nodes = 40
l3_nodes = 30 
output_nodes = 10

number_of_parameters = (input_nodes*l1_nodes + l1_nodes) + (l1_nodes*l2_nodes + l2_nodes) + (l2_nodes*l3_nodes + l3_nodes) + (l3_nodes*output_nodes + output_nodes)

print("The number of parameters in network (a)  = ", number_of_parameters)


The number of parameters in network (a)  =  822830


The number of parameters in network (a) = 822830

(b)

The size of feature map can be calculated with the formula **feature_map_size = (n+2p-f)/s +1** ;where n is the size of the image, p is the padding, f is the size of the filter and s is the stride. In this network, p=0, s=1 (without stride), f=8, and n=128. Then, feature_map_size = (128-8) +1 = 121. So, the feature map has size 121 x 121. 

(c)

By applying s=4 into the formular in question (b), feature_map_size = [(128-8)/4] +1 = 31. So, the feature map has size 31 x 31.


(d)

Without using weigh sharing, each feature in the feature maps learn different weights and biases.

* A 8x8 filter between input and first hidden
layer (with stride = 4 and a different bias for each hidden node) has number of parameters **p1 = parameter_for_each_feature_map X size_of_feature_map**. 

  We have :
  * parameter_for_each_feature_map = 8x8 + 1 = 65
  * size_of_feature_map = [(128-8)/4 +1] = 31

  => p1 = 65 x 31 x 31 = 62465

* A 8x8 filter between first and second hidden layer (with stride = 4 and a different bias for each hidden node) has number of parameters **p2 = parameter_for_each_feature_map X size_of_feature_map**.

  We have :
  * parameter_for_each_feature_map = 8x8 + 1 = 65
  * size_of_feature_map = [int((31-8)/4) +1] = 6

  => p2 = 65 x 6 x 6 = 2340

* A fully connected output layer has the parameter p3 = size_of_feature_map x size_of_feature_map x 10 + 10 = 6 x 6 x 10 + 10 = 370

**The total number of parameters p = p1 + p2 + p3 = 62450 + 2340 + 370 = 65160**.





(e) 

With weight sharing, the network is the convolutional network. 

*   A 8x8 filter between input and first hidden layer (with stride = 4 and a shared bias for each feature map) has number of parameters p1 = 8x8 +1 = 65. 
*   A 8x8 filter between first and second hidden layer (with stride = 4 and a shared bias for each feature map) has number of parameters p2 = 8x8 +1 = 65.
*   A fully connected output layer has number of parameters p3 = size_of_feature_map x size_of_feature_map x 10 + 10 = 6 x 6 x 10 + 10 = 370

**The total number of parameters p = p1 + p2 + p3 = 65 + 65 + 370 = 500**.

